### ISSUE TITLE
Azure SQL Publish not supported on Linux build agent

### ISSUE THREAD

**Issue by severisv** [Microsoft/azure-pipelines-tasks#8408]
Azure SQL Publish build task is not supported on agent Hosted Linux Preview:

>The current operating system is not capable of running this task. That typically means the task was written for Windows only. For example, written for Windows Desktop PowerShell.

Are there any other options available that will allow me to migrate my Azure SQL database during deployment?

**Comment by Ajay-MS** [comment]
@severisv 

I would like to understand your scenario before suggesting a solution to the problem.  Please share your email Id so that we can plan for a call to look into this issue. 

Few basic questions : 

-  In what format your source data exists. 
-  How are you migrating your data to the Azure SQL database?

**Comment by severisv** [comment]
We are using Entity Framework Core. Previously we would have `dotnet ef` create an SQL script that we would run agaist the Database through the Azure SQL Publish task.

After changing to a Hosted Linux Preview agent, we no longer have a way of running the SQL script against the database during deployment.

**Comment by Ajay-MS** [comment]
@severisv 

Azure SQL task is currently supported only for Windows-based agent system. We are adding this to our backlog and will consider adding this support in the future. 

I suggest to try out below-mentioned solutions as of now:
1. Using bash script task,
       -  Install MS SQL tools on linux agent
       -  Use bash script task to run script.  
       -  Reference - [Create and run SQL Server Agent jobs on Linux](https://docs.microsoft.com/en-us/sql/linux/sql-server-linux-run-sql-server-agent-job?view=sql-server-2017)

2.  You can use Azure RG task and execute store proc as part of the ARM template 
      Reference - [Copy data from Azure Blob Storage to Azure SQL Db with Stored Procedure](https://github.com/Azure/azure-quickstart-templates/tree/master/101-data-factory-blob-to-sql-copy-stored-proc)

I hope this would be helpful.

**Comment by Ajay-MS** [comment]
@severisv 

Closing this issue assuming you are unblocked on this issue.

**Comment by Akreddy123** [comment]
Facing the same issue with below error.
**#[error]The current operating system is not capable of running this task. That typically means the task was written for Windows only. For example, written for Windows Desktop PowerShell.**

Can you please tell me what is wrong with my task?

**Comment by Ajay-MS** [comment]
@Akreddy123 

Generally this issue occurs, if you are using PS based task or Node based task with dependency on windows platform tool and running the task on Linux agent. 

Can you please provide following information :
1. Are you using any pre-defined task or have you created a new task. 
2. If  pre-defined task, can you share the name of the task?
3. If you have created the task, which handler for the task you are using - Node or PS ?
4. On which platform your agent is running - Linux or Windows ?

**Comment by Akreddy123** [comment]
1. We have created an extension named "Secure DevOps Kit (AzSK) CICD Extensions for Azure" with 2 tasks in it
  i. AzSKSVTs
  ii .AzSKARMTemplateChecker
2. We are using a PS handler.
3. Both the tasks are running on Windows agent but giving the above error on Linux agent.
Note : The commands which we use in our tasks are completely working fine in Linux machine.

**Comment by kmkumaran** [comment]
PS handler works only in Windows. If you need support in Linux, you will have to use Node handler. You can refer the PS task that we have as an example:
https://github.com/microsoft/azure-pipelines-tasks/blob/master/Tasks/PowerShellV2/task.json

**Comment by Akreddy123** [comment]
@kmkumaran : In the above task, you are using 2 separate targets for power shell and node. In our task, we use some commands of Az modules and our AzSK module. 
Is there a way other than using 2 separate targets?

**Comment by kmkumaran** [comment]
If you need only one handler, you can then go with just Node handler. The above task is an example where task has an option to specify two handlers and the best handler will be picked based on operating system. But if you always want to use Node, you can then have only Node handler. The same repo has other tasks which uses only Node handler. You can also refer them as examples.

**Comment by Akreddy123** [comment]
I have to use the PS handler since I use PS commands in my task. 
Is there a way other than using Node to make it work on Linux agent?

**Comment by kmkumaran** [comment]
Today, the only handler we support for Linux is Node. You can use [Powershell core](https://docs.microsoft.com/en-us/powershell/scripting/whats-new/what-s-new-in-powershell-core-60?view=powershell-6), which can let you execute powershell commands in both Windows and Linux. You can then use Node handler and follow similar approach as in the [Powershell task](https://github.com/microsoft/azure-pipelines-tasks/blob/master/Tasks/PowerShellV2/powershell.ts) to run your scripts (ps commands) in Powershell core.  

Adding @ericsciple if there are any other suggestions.

**Comment by Akreddy123** [comment]
@kmkumaran : We are using PowerShell core only. That's why the commands we use in the above task are running on Linux and Mac VMs.
My question is what should I do to make my task run on Hosted Ubuntu or any other Linux agent. I can't see any azure PowerShell command in your type script file. Please let me know if it is possible to run any azure PowerShell command in a type script file or is there a way other than using Node.

**Comment by Akreddy123** [comment]
@kmkumaran @ericsciple @Ajay-MS : Could you please respond to my comment.

**Comment by kmkumaran** [comment]
@Akreddy123 - I believe having Node handler is the only option for you. With regard to your observation of PowerShell command in the type script, Please note that this task takes a PS script (inline or file) as user input and executes them. Hence you don't see the commands in the task file. But the functionality that it does - to be able to execute a PS script using pwsh would be of interest to you.

**Comment by rosieks** [comment]
Is there any plan to provide SQL task that works on Linux too or at least instruction how to do it? Unfortunatelly instruction like that:

>- Install MS SQL tools on linux agent
>- Use bash script task to run script.
Is quite short.

**Comment by DmytroSokhach** [comment]
This link has pretty good manual: [https://www.programmingwithwolfgang.com/deploy-dacpac-linux-azure-devops/](url) "Automatically Deploy your Database with Dacpac Packages using Linux and Azure DevOps"

**Comment by k290** [comment]
Fixing the link Dmytro posted. When you click it it goes to the wrong URL
https://www.programmingwithwolfgang.com/deploy-dacpac-linux-azure-devops/

**Comment by k290** [comment]
Since people continue to post relatively difficult to figure out and implement workarounds in this thread I thought it might be worthwhile to open a feature request for people to vote on:
https://developercommunity.visualstudio.com/t/Allow-Azure-SQL-Publish-SqlAzureDacpacD/1453203
Please cast your vote if you think it would be worthwhile

**Comment by pawel-zolty** [comment]
5 years and there is still no task for deploying sql database from linux agent?

**Comment by vanmash** [comment]
[Microsoft loves Linux](https://cloudblogs.microsoft.com/windowsserver/2015/05/06/microsoft-loves-linux/), really?

Please support this task on Linux

**Comment by slnali** [comment]
workaround using azure devops in 2 lines: 

        - pwsh: sudo dotnet tool install -g microsoft.sqlpackage
        - pwsh: sqlpackage /Action:Publish /SourceFile:"$(Pipeline.Workspace)/$(artifactName)/{{YOUR DACPACK FILE NAME}}.dacpac" /TargetConnectionString:"$(sql_connectionString)"


https://learn.microsoft.com/en-us/sql/tools/sqlpackage/sqlpackage-publish?view=sql-server-ver16

**Comment by VaskoBoshkov** [comment]
> workaround using azure devops in 2 lines:
> 
> ```
>     - pwsh: sudo dotnet tool install -g microsoft.sqlpackage
>     - pwsh: sqlpackage /Action:Publish /SourceFile:"$(Pipeline.Workspace)/$(artifactName)/{{YOUR DACPACK FILE NAME}}.dacpac" /TargetConnectionString:"$(sql_connectionString)"
> ```
> 
> https://learn.microsoft.com/en-us/sql/tools/sqlpackage/sqlpackage-publish?view=sql-server-ver16

But, how to create the dacpac file?

**Comment by rayou18** [comment]
I looked into the SqlAzureDacpacDeployment@1 task and in their readme they mention the usage of the Invoke-Sqlcmd cmdlet. Instead of the dacpac file you can use a migration file/.sql file.

1. install powershell on your linux agent
2. install the sqlserver module:  Install-Module sqlserver
3. run the following command using the Invoke-Sqlcmd cmdlet using the PowerShell@2 with the pwsh input set to true:
  - task: PowerShell@2
      inputs:
        targetType: 'inline'
        script: |
            Invoke-Sqlcmd -InputFile <reference to your .sql file> ConnectionString '<your connection string>'
        pwsh: true # This runs the script with PowerShell Core on Linux.

**Comment by k290** [comment]
> I looked into the SqlAzureDacpacDeployment@1 task and in their readme they mention the usage of the Invoke-Sqlcmd cmdlet. Instead of the dacpac file you can use a migration file/.sql file.
> 
> 1. install powershell on your linux agent
> 2. install the sqlserver module:  Install-Module sqlserver
> 3. run the following command using the Invoke-Sqlcmd cmdlet using the PowerShell@2 task with the pwsh input set to true:
> 
> ```
>   - task: PowerShell@2
>       inputs:
>         targetType: 'inline'
>         script: |
>             Invoke-Sqlcmd -InputFile <reference to your .sql file> ConnectionString '<your connection string>'
>         pwsh: true # This runs the script with PowerShell Core on Linux.
> ```

Useful for those with their own agents. But for microsoft hosted agents its still a problem

**Comment by VaskoBoshkov** [comment]
> I looked into the SqlAzureDacpacDeployment@1 task and in their readme they mention the usage of the Invoke-Sqlcmd cmdlet. Instead of the dacpac file you can use a migration file/.sql file.
> 
> 1. install powershell on your linux agent
> 2. install the sqlserver module:  Install-Module sqlserver
> 3. run the following command using the Invoke-Sqlcmd cmdlet using the PowerShell@2 task with the pwsh input set to true:
> 
> ```
>   - task: PowerShell@2
>       inputs:
>         targetType: 'inline'
>         script: |
>             Invoke-Sqlcmd -InputFile <reference to your .sql file> ConnectionString '<your connection string>'
>         pwsh: true # This runs the script with PowerShell Core on Linux.
> ```

Well, this only works for SQL Server. What about Postgresql?

**Comment by VaskoBoshkov** [comment]
Okay, i figured it out for PostgreSQL:

psql -d "dbname='XXX' user='XXX' password='XXX' host='XXX'" --single-transaction -v ON_ERROR_STOP=on -f XXX.sql

**Comment by marecki9** [comment]
The challenge I faced was connecting to the database using a service principal account. I invested significant time experimenting with various Azure cmdlets, trying to understand why my build agent couldn't access the desired subscription.

Then I discovered a task that automated the connection process to the Azure. This eliminated the need for calling `Clear-AzContext/Connect-AzAccount/Set-AzContext`, making the entire operation a ridiculously simple thing. See the example:

```
- task: AzurePowerShell@5
  displayName: Azure DB Command
  inputs:
    azureSubscription: '$(azureSubscription)'
    ScriptType: 'InlineScript'
    Inline: |
      $accessToken = (Get-AzAccessToken -ResourceUrl https://database.windows.net).Token

      Invoke-SqlCmd -ServerInstance $(sqlServer) `
                    -Database $(databaseName) `
                    -AccessToken $accessToken `
                    -Query "<your query here>"
    FailOnStandardError: true
    azurePowerShellVersion: 'LatestVersion'
    pwsh: true
```

**Comment by marecki9** [comment]
I have also updated the example above to deploy the database using `.dacpac` and `.publish.xml` files. This approach was applied in my release pipeline, where I was using the `SqlAzureDacpacDeployment@1` task. The goal was to switch to a Linux agent while maintaining this deployment strategy. The only adjustments are needed in the `Inline` section of the task.

**1. Required tool:**

Ensure the `microsoft.sqlpackage` .NET tool is installed on your agent. If it's not installed, run the following command to install it globally:
```pwsh
dotnet tool install "microsoft.sqlpackage" -g
```


**2. Enviroment paths**

To ensure global tools are recognized on your Linux agent, add the path to your environment. If not, `sqlpackage` might not be recognized as a valid command. This can be handled using PowerShell:

```pwsh
$dotnetToolsPath = "$HOME/.dotnet/tools"

if ($env:PATH -notlike "*$dotnetToolsPath*") {
    $env:PATH += ":$dotnetToolsPath"
}
```

**3. Locate your dacpac & publish files**

Use the following script to set the correct paths for your .dacpac and .publish.xml files:

```pwsh
$rootDirectory = "$(System.DefaultWorkingDirectory)/<path to an artifact or build folder>"
$dacPacFilePath = "$($rootDirectory)/<database project>.dacpac"
$publishProfileFilePath = "$($rootDirectory)/<database project>.publish.xml"
```

**4. Build your connection string**

- SQL Server Authentication

Use the standard Server/Database/User Id/Password combination in your connection string.

- Service Principal

To authenticate using a service principal, you’ll need to obtain an access token instead of User Id and Password. You can retrieve it as follows:

```pwsh 
$accessToken = (Get-AzAccessToken -ResourceUrl "https://database.windows.net").Token
```

_Note: [Az.Accounts](https://learn.microsoft.com/en-us/powershell/module/az.accounts/?view=azps-12.2.0) PowerShell module is required to use it!_

**5. Deploy**

- SQL Server Authentication:

Use the following command to deploy your database:
```pwsh
sqlpackage `
  /Action:publish `
  /Sourcefile:$dacPacFilePath `
  /Profile:$publishProfileFilePath `
  /TargetConnectionString:$connectionString
```

- Service Principal:

For service principal-based authentication, include the access token:
```pwsh
sqlpackage `
  /at:$accessToken `
  /Action:publish `
  /Sourcefile:$dacPacFilePath `
  /Profile:$publishProfileFilePath `
  /TargetConnectionString:$connectionString
```
